platform-sdk/docs/proposals/001-merkle-state-api/001-merkle-state-api.md

```diff
@@ -0,0 +1,300 @@
+# Merkle State API
+
+## Summary
+
+Refactor the Merkle tree so that it can be fully represented by the State API objects.
+Create a module that will provide a simple and easy-to-use API for the Merkle tree.
+
+## Purpose and Context
+
+### Why?
+
+The Merkle tree is a rather complex data structure, and to make interaction easier, we need to hide this complexity behind the API.
+Another reason to have this API is to provide additional flexibility in changing the implementation details of the Merkle tree.
+As long as they are hidden, we are free to change them.
+
+Note that this API is already implemented as a part of the Service Modularization project, but these classes will be moved
+to a separate module as a part of this proposal.
+
+### Context
+
+There are two issues that need to be addressed:
+
+* We need to have a public API for the Merkle tree that is simple and easy to use, residing in a separate module with a minimal set of dependencies.
+* The Merkle tree has to be fully expressed in terms of State API. Currently, `PlatformState` stands out and needs to be refactored.
+
+#### Singleton
+
+A singleton is a node with 2 children:
+
+- string label for State Name
+- protobuf message for child data
+
+In Java code it's represented by the following classes:
+```java
+public interface ReadableSingletonState<T> {
+  @NonNull
+  String getStateKey();
+  T get();
+  boolean isRead();
+}
+
+public interface WritableSingletonState<T> extends ReadableSingletonState<T> {
+  void put(@Nullable T value);
+
+  boolean isModified();
+}
+```
+
+## Requirements
+
+- `PlatformState` should be refactored to a singleton object as defined by the State API.
+- there should be a new module - `swirlds-state-api` - to host a set of interfaces, records, and abstract classes that represent the Hedera state.
```

Comment:
```suggestion
- there should be a new module - `swirlds-state-api` - to host a set of interfaces, records, and abstract classes that represent the Hashgraph state.
```

Reply:
I honestly don't know what to call it. In the open source world `Swirlds` isn't right, and `Hedera` isn't right, and `Hashgraph` might be right but also might be misleading because it makes me think that the hashgraph itself is stored in state, but it is not.

Reply:
@rbair23 would `ApplicationState` or `AppState` be too generic?

Reply:
As long as people agree, I can go with `MerkleState`. I don't have a strong opinion

Reply:
"Ledger State" is also an option. I don't oppose "Merkle State" though.

Reply:
The majority agrees on `MerkleState`. So, let it be it.